### PR TITLE
Implemented modifyTable in BigtableAdmin with its test cases.

### PR DESCRIPTION
**Pull request by rupeshitpatekarq**
@sduskis Done with one type of modifyTable implementation, could you please have a look on it.

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestModifyTable.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      Put put = new Put(rowKey);
+      put.addColumn(COLUMN_FAMILY, qual, 1L, value);
+      put.addColumn(COLUMN_FAMILY, qual, 2L, value);
+      table.put(put);
+      assertTrue(table.exists(new Get(rowKey)));
+      HTableDescriptor descriptor = new HTableDescriptor(tableName);
+      descriptor.addFamily(new HColumnDescriptor(COLUMN_FAMILY2));
+      modifyTable(descriptor);
+      assertTrue(admin.tableExists(tableName));
+      assertTrue(admin.getTableDescriptor(tableName).hasFamily(COLUMN_FAMILY2));
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestModifyTable.java]: Please add a test to make sure that the table descriptor has no other columns.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestModifyTable.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  
+  public static final byte[] COLUMN_FAMILY2 = Bytes.toBytes("test_family2");
+  
+  @Test
+  public void testModifyTable() throws IOException {
+    TableName tableName = sharedTestEnv.newTestTableName();
+    sharedTestEnv.createTable(tableName);
+    try(Admin admin = getConnection().getAdmin();
+        Table table = getConnection().getTable(tableName)) {
+      byte[] rowKey = dataHelper.randomData("testrow-");
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestModifyTable.java]: Please remove any testing of data, including `Put`s.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 3]
```diff
@@ -360,15 +362,82 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
+  public void modifyTable(TableDescriptor tableDescriptor) throws IOException {
+    modifyTable(tableDescriptor.getTableName(), tableDescriptor);
   }
 
   @Override
-  public void modifyTable(TableName arg0, TableDescriptor arg1) throws IOException {
-    throw new UnsupportedOperationException("modifyTable"); // TODO
+  public void modifyTable(TableName tableName, TableDescriptor tableDescriptor) throws IOException {
+    //Check table exists or not
+    if(isTableAvailable(tableName)) {
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please get the current descriptor here.  Once you have that, we need to create an array of modifications that need to occur.  Those modifications need to be sent in a single RPC.

There's a method in AbstractBigtableAdmin:

```
protected void modifyColumn(TableName tableName, String columnName,
      String modificationType, Modification... modifications) 
```
Please use this method with an array of modifications.  That will send a single RPC.

**Comment by rupeshitpatekarq** [comment]
@sduskis, I have addressed all of the comments. Could you please take a look.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 4]
```diff
@@ -360,15 +361,78 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
+    modifyTable(tableDescriptor.getTableName(), tableDescriptor);
   }
 
   @Override
-  public void modifyTable(TableName arg0, TableDescriptor arg1) throws IOException {
-    throw new UnsupportedOperationException("modifyTable"); // TODO
+  public void modifyTable(TableName tableName, TableDescriptor tableDescriptor) throws IOException {
+    if(isTableAvailable(tableName)) {
+      List<Modification> modifications = new ArrayList<>();
+      buildModificationsForModifyFamilies(tableName,tableDescriptor,modifications);
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please add a space after `,`'s here and the next couple of lines.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 5]
```diff
@@ -360,15 +361,78 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
 
+  /**
+   * This method will build list of either create or modify Modifications objects..
+   * 
+   * @param tableName
+   * @param tableDescriptor
+   * @param modifications an array of {@link com.google.bigtable.admin.v2.ModifyColumnFamiliesRequest.Modification} object.
+   * @throws IOException
+   */
+  private void buildModificationsForModifyFamilies(TableName tableName, TableDescriptor tableDescriptor,List<Modification> modifications) throws IOException {
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: please add a space after ','

Comment (sduskis) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Also, this code should be in the `bigtable-hbase` project, and ideally a stand alone class.  This logic needs to be used for the HBase 1 `BigtableAdmin` and also the HBase 2 `AsyncAdmin`.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 6]
```diff
@@ -360,15 +361,78 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
+  
+  /**
+   * This method will build list of delete Modifications objects. 
+   * 
+   * @param tableName
+   * @param tableDescriptor
+   * @param modifications an array of {@link com.google.bigtable.admin.v2.ModifyColumnFamiliesRequest.Modification} object.
+   * @throws IOException
+   */
+  private void buildModifcationsForDeleteFamilies(TableName tableName, TableDescriptor tableDescriptor,List<Modification> modifications) throws IOException {
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: same here.. this ought to be reusable for hbase1, hbase2 and hbase 2 async.

**Comment by rupeshitpatekarq** [comment]
@sduskis, Created a stand-alone class AbstractBigtableModifyTable which can be reused for hbase1, hbase2 and hbase2 async.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestModifyTable.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import org.junit.runners.JUnit4;
+
+@SuppressWarnings("deprecation")
+@RunWith(JUnit4.class)
+public class TestModifyTable extends AbstractTestModifyTable {
+
+  @Override
+  protected void modifyTable(HTableDescriptor descriptor) throws IOException {
+    try(Admin admin = getConnection().getAdmin()) {
+      admin.modifyTable(descriptor);
```
[the remark this answers is not in the document]
Comment (rupeshitpatekarq) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestModifyTable.java]: We can't get TableDescriptor in AbstractTestModifyTable hence I have used HTableDescriptor.

[the remark this answers is not in the document]
Comment (sduskis) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestModifyTable.java]: You'll need to cast the `HTableDescriptor` to a `TableDescriptor` to force the use of the new `TableDescriptor` based method.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 8]
```diff
@@ -360,15 +361,34 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
-    throw new UnsupportedOperationException("modifyTable"); // TODO
+  public void modifyTable(TableDescriptor tableDescriptor) throws IOException {
+    modifyTable(tableDescriptor.getTableName(), tableDescriptor);
   }
 
   @Override
-  public void modifyTable(TableName arg0, TableDescriptor arg1) throws IOException {
-    throw new UnsupportedOperationException("modifyTable"); // TODO
+  public void modifyTable(TableName tableName, TableDescriptor tableDescriptor) throws IOException {
+    System.out.println("********************* IN modifyTable ***********************");
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please remove the print statement.

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 9]
```diff
@@ -360,15 +361,34 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
+    if (isTableAvailable(tableName)) {
+      TableDescriptor currentTableDescriptor = getTableDescriptor(tableName);
+      List<Modification> modifications = new ArrayList<>();
+      List<HColumnDescriptor> columnDescriptors = toHColumnDescriptors(tableDescriptor);
+      List<HColumnDescriptor> currentColumnDescriptors = toHColumnDescriptors(currentTableDescriptor);
+      modifications.addAll(tableModificationAdapter.buildModificationsForModifyFamilies(columnDescriptors, currentColumnDescriptors));
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please combine the two calls into the adapter into a single one that performs the same ling:  `tableModificationAdapter.buildModifications(columnDescriptors, currentColumnDescriptors)`

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 10]
```diff
@@ -360,15 +361,34 @@ public void deleteColumnFamily(TableName tableName, byte[] columnName) throws IO
...
+      modifications.addAll(tableModificationAdapter.buildModifcationsForDeleteFamilies(columnDescriptors, currentColumnDescriptors));
+      modifyColumn(tableName, "modifyTable", "update", (Modification[]) modifications.toArray());
+    } else {
+      throw new TableNotFoundException(tableName);
+    }
   }
 
+  private List<HColumnDescriptor> toHColumnDescriptors(TableDescriptor tableDescriptor) {
```
Comment (sduskis) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please move this logic into `tableAdapter2x`.

**Comment by rupeshitpatekarq** [comment]
@sduskis, Addressed all comments.

**Pull request merged**